lua/neogit/status.lua

```diff
@@ -64,391 +70,245 @@ local function get_section_item_for_line(linenr)
-  status_buffer:set_foldlevel(2)
+--- Find the closest section the cursor is encosed by.
+--
+-- @return table, table, table, number, number -
+--  The first 3 values are tables in the shape of {number, string}, where the number is
+--  the relative offset of the found item and the string is it's identifier.
+--  The remaining 2 numbers are the first and last line of the found section.
+local function save_cursor_location()
+  local line = vim.fn.line('.')
+  local section_loc, file_loc, hunk_loc, first, last
 
-  for _,l in pairs(locations) do
-    local items = status[l.name]
-    if items ~= nil and l.name ~= "stashes" and l.name ~= "unpulled" and l.name ~= "unmerged" then
-      for _, i in pairs(items) do
-        if i.diff_content ~= nil then
-          for _,h in ipairs(i.diff_content.hunks) do
-            status_buffer:create_fold(i.first + h.first, i.first + h.last)
-            status_buffer:open_fold(i.first + h.first)
-            if not config.values.disable_signs then
-              status_buffer:place_sign(i.first + h.first, "NeogitOpen:hunk", "fold_markers")
+  for li, loc in ipairs(locations) do
+    if line == loc.first then
+      section_loc = {li, loc.name}
+      first, last = loc.first, loc.last
+      break
+    elseif line >= loc.first and line <= loc.last then
+      section_loc = {li, loc.name}
+      for fi, file in ipairs(loc.files) do
+        if line == file.first then
+          file_loc = {fi, file.name}
+          first, last = file.first, file.last
+          break
+        elseif line >= file.first and line <= file.last then
+          file_loc = {fi, file.name}
+          for hi, hunk in ipairs(file.hunks) do
+            if line <= hunk.last then
+              hunk_loc = {hi, hunk.hash}
+              first, last = hunk.first, hunk.last
+              break
             end
           end
-        end
-        status_buffer:create_fold(i.first, i.last)
-        if not config.values.disable_signs and l.name ~= "untracked_files" then
-          status_buffer:place_sign(i.first, "NeogitClosed:item", "fold_markers")
+          break
         end
       end
-    end
-    status_buffer:create_fold(l.first, l.last)
-    status_buffer:open_fold(l.first)
-    if not config.values.disable_signs then
-      status_buffer:place_sign(l.first, "NeogitOpen:section", "fold_markers")
+      break
     end
   end
 
-  vim.fn.winrestview(old_view)
-  status_buffer:lock()
+  return section_loc, file_loc, hunk_loc, first, last
 end
 
-function primitive_move_cursor(line)
-  for _,l in pairs(locations) do
-    if l.first <= line and line <= l.last then
-      vim.api.nvim_win_set_cursor(0, { l.first, 0 })
-      break
-    end
+local function restore_cursor_location(section_loc, file_loc, hunk_loc)
+  if #locations == 0 then return vim.fn.setpos('.', {0, 1, 0, 0}) end
+  if not section_loc then section_loc = {1, ''} end
+
+  local section = Collection.new(locations):find(F.C(F.dot('name'), F.eq(section_loc[2])))
```

Comment:
@TimUntersberger raising this one myself, since I know this can raise an eyebrow or 2 :sweat_smile:
Am I overdoing it here? I'm used to thinking this way and thus tend to solve problems through function composition; so I thought I'd ask with a trivial problem (the composition can easily be replaced by `function (i) return i.name == section_loc[2] end`) if code like this is alright with you.

Reply:
I am a fan of functional programming, but most of the time it overcomplicates simple things just like in this case imo.

Reply:
I do agree on the sentiment, but I feel *overcomplicated* is the wrong way to put it.
For example, given fsharp where I can easily express the same logic in both paradigms:
```fsharp
dot "name" >> eq section[2]
fun x -> x.name == section[2]
```
both look equally complex to me.

Lua is not nearly as expressive though, so I need to add a ton of **noise** to represent function composition. So, while writing the function out normally is actually *longer* than using my composition helper, it looks cleaner, since more of the elements used convey useful meaning.

I'll put in normal functions there, but I think I'll keep the composition function around; since you're up for it in principle, it will be useful some other time I hope :grin: